src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -181,142 +187,184 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
+                  || currentChar == '\r'
+                  || Math.min(movingEndByte, maxByteRange) == maxByteRange) {
+                newLineFound = true;
+                partitionInputStream =
+                    new SequenceInputStream(
+                        partitionInputStream,
+                        source.getInputStream(
+                            adjustedEndByteRange + 1, Math.min(movingEndByte, maxByteRange)));
+                BufferedReader reader =
+                    new BufferedReader(new InputStreamReader(partitionInputStream));
+                reader.skip(characterPositionAtBeginningOfRecord);
+                parser =
+                    new CSVParser(
+                        reader,
+                        CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
+                        0,
+                        0);
+                iterator = parser.iterator();
+                record = iterator.next();
+                finishedReadingLastRow = true;
+                break;
+              }
+              dataChar = trailingEndInputStream.read();
+            }
+            trailingEndInputStream.close();
           }
-          partitionEndByteRange += byteOverlap;
+        } else if (record.size() == schema.numColumns() && onLastRow && isLastWorker) {
+          /*
+           * If we are on the last worker, just mark it as finished.
+           */
+          finishedReadingLastRow = true;
+        }
+
+        /*
+         * This is the case when we are missing data for a worker. We simply request more bytes into the stream.
+         * Although we are on the last row at this point, we don't mark it as finished unless we reach the max size of
+         * the file
+         */
+        if (record.size() < schema.numColumns() && !isLastWorker) {
+          long characterPositionAtBeginningOfRecord = record.getCharacterPosition();
           InputStream overlapStream =
-              source.getInputStream(partitionStartByteRange, partitionEndByteRange);
+              source.getInputStream(
+                  adjustedEndByteRange + 1,
+                  Math.min(adjustedEndByteRange + byteOverlap, maxByteRange));
           partitionInputStream = new SequenceInputStream(partitionInputStream, overlapStream);
+          BufferedReader reader = new BufferedReader(new InputStreamReader(partitionInputStream));
+          reader.skip(characterPositionAtBeginningOfRecord);
           parser =
               new CSVParser(
-                  new BufferedReader(new InputStreamReader(partitionInputStream)),
-                  CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape));
+                  reader, CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape), 0, 0);
           iterator = parser.iterator();
+          record = iterator.next();
           byteOverlap *= 2;
-        } else {
-          discardedRecord = true;
+          adjustedEndByteRange += byteOverlap;
```

Comment:
Are you sure you want to increment `adjustedEndByteRange` by `byteOverlap` before making the comparison in the next line?